### PR TITLE
GNAF Dataset to Dataframe

### PR DESCRIPTION
**Pull request by zazwaz12**
Please make review this code for logic errors, or redundant code (within the same file for now - as we are not orchestrating yet).

### CODE REVIEW

**Review by ctmes — APPROVED** [review]

File: 1. extraction/gnaf.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+# Accessing the DataFiles Path
+home_dir = os.path.expanduser("~")
+datafiles_path = os.path.join(home_dir, "Desktop", "DataFiles")
+
+# Setting the FilesIn as the source files, and the AppStaging for the pickles
+in_path = os.path.join(datafiles_path, "FilesIn")
+standard_in = os.path.join(in_path, "Standard")
+authority_code_in = os.path.join(in_path, "Authority Code")
+pickle_file_path = os.path.join(datafiles_path, "AppStaging", 'locational-data.pkl')
+files_out_path = os.path.join(datafiles_path, "FilesOut")
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:14]: Please don't hard code in the file paths, have these as inputs to the functions instead. Also, avoid using global variables

File: 1. extraction/gnaf.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+def read_codes():
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:16]: docstring of what the function do would be nice

File: 1. extraction/gnaf.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                key = os.path.splitext(filename)[0]
+                dataframes_dict[key] = df
+                
+            # The code files may be empty - the version I had were empty
+            except pd.errors.EmptyDataError:
+                continue
+            except pd.errors.ParserError:
+                print(f"Skipped malformed file: {filename}")
+            except Exception as e:
+                print(f"Error processing file {filename}: {e}")
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:39]: use `logger.warning()` instead of `print`, also you can just wrap the function in `@logger.catch()` to catch and warn about the exceptions without terminating

File: 1. extraction/gnaf.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Write the dictionary of DataFrames to an Excel file
+    with pd.ExcelWriter(os.path.join(files_out_path, 'code-meanings.xlsx')) as writer:
+        for sheet_name, df in dataframes_dict.items():
+            df.to_excel(writer, sheet_name=sheet_name, index=False)
+
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:45]: what's the purpose of this?

File: 1. extraction/gnaf.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        if filename.endswith('.psv'):
+            file_path = os.path.join(standard_in, filename)
+            try:
+                df = pd.read_csv(file_path, delimiter='|')
+                key = os.path.splitext(filename)[0]
+                dataframes_dict[key] = df
+            except pd.errors.EmptyDataError:
+                print(f"Skipped empty file: {filename}")
+            except Exception as e:
+                print(f"Error processing file {filename}: {e}")
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:62]: duplicate code with those in `read_codes`

File: 1. extraction/gnaf.py (hunk 3 above)
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:39]: use `logger.warning()` instead of `print`, also you can just wrap the function in `@logger.catch()` to catch and warn about the exceptions without terminating

File: 1. extraction/gnaf.py (hunk 2 above)
Comment (SodaVolcano) [reply 1. extraction/gnaf.py:16]: Also, `read_codes` is not too descriptive - what "code" refers to is unclear

File: 1. extraction/gnaf.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    with open(pickle_file_path, 'wb') as pickle_file:
+        pickle.dump(dataframes_dict, pickle_file)
+
+def output_files_summaries():
+    print("Reading in the pickle")
+    with open(pickle_file_path, 'rb') as file:
+        data_dict = pickle.load(file)
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:70]: path of pickle should be an argument to the function

File: 1. extraction/gnaf.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            df_shape = df.shape
+            df_head = df.head().to_string(index=False)
+            total_rows += df_shape[0]
+            data_info.append([file_name, df_shape[1], df_shape[0], df_head])
+            filtered_data_dict[file_name] = df
+
+    data_info.append(["All these files put together are", '', total_rows, ''])
+    info_df = pd.DataFrame(data_info, columns=['File Name', 'Columns', 'Rows', 'Head'])
+    info_df.to_csv(os.path.join(files_out_path,'data_house_info.csv'), index=False)
+    return filtered_data_dict
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:102]: These functions seem to be one-off operations, try make your functions reusable and documented

File: 1. extraction/gnaf.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+if __name__ == "__main__":
+    # Run the functions for Authority Code and Standard files
+    #read_codes()
+    #read_data_to_dictionary_pickle()
+    data_dict = output_files_summaries()
+    filter_by_lat_lon(data_dict)
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:109]: shouldn't have this part for a module

File: 1. extraction/gnaf.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+
+# Example logic check (to be transitioned into unit tests)
+"""
+print(len(os.listdir(standard_in)))  # LOGIC CHECK SHOULD EQUAL 171
+print(len(os.listdir(authority_code_in)))  # LOGIC CHECK SHOULD EQUAL 16
+"""
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:116]: use `assert` instead, and add to `pytest`

File: syncdata.py
Comment (SodaVolcano) [inline syncdata.py:1]: why was this deleted?

File: 1. extraction/gnaf.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
+import pandas as pd
```
Comment (SodaVolcano) [inline 1. extraction/gnaf.py:1]: Please use polars instead of pandas

File: 2. transformation/houses.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import os
+
+# Accessing the DataFiles Path
+home_dir = os.path.expanduser("~")
+datafiles_path = os.path.join(home_dir, "Desktop", "DataFiles")
+
+# Setting the FilesOut and the AppStaging for the pickles
+pickle_file_path = os.path.join(datafiles_path, "AppStaging", 'locational-data.pkl')
+files_out_path = os.path.join(datafiles_path, "FilesOut")
+output_pickle_path = os.path.join(datafiles_path, "AppStaging", 'housing_data.pkl')
```
Comment (SodaVolcano) [inline 2. transformation/houses.py:16]: same comments for global variables as above

File: 2. transformation/houses.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    print(concatenated_df)
+    # Saving the concatenated DataFrame to a pickle file
+    with open(output_pickle_path, 'wb') as file:
+        pickle.dump(concatenated_df, file)
+
+    print(f"Concatenated DataFrame saved to {output_pickle_path}")
+
+    # Sampling 750,000 rows from the concat DF for future reference
+    sample_df = concatenated_df.sample(n=750000, random_state=1)
+    sample_df.to_csv(os.path.join(files_out_path, "houses-sample.csv"), index=False)
```
Comment (SodaVolcano) [inline 2. transformation/houses.py:45]: the read_XYZ functions can be generalised to work with any folders, don't have to be specific e.g. read_home, read_codes, read_standards

**Review by SodaVolcano — CHANGES_REQUESTED** [review]
Sorry I don't think the code should be merged, my comments are below. I've written an equivalent version in a separate branch which I will push soon for you to compare

**Comment by ctmes** [comment]
> Sorry I don't think the code should be merged, my comments are below. I've written an equivalent version in a separate branch which I will push soon for you to compare

This is mainly what my auto testing branch is for right?

File: 1. extraction/gnaf.py (hunk 1 above)
Comment (zazwaz12) [reply 1. extraction/gnaf.py:14]: This is code that will only be run once.

File: 1. extraction/gnaf.py (hunk 2 above)
Comment (zazwaz12) [reply 1. extraction/gnaf.py:16]: It reads the authority codes.

File: 1. extraction/gnaf.py (hunk 3 above)
Comment (zazwaz12) [reply 1. extraction/gnaf.py:39]: Happy to add that, functionality first.

File: 1. extraction/gnaf.py (hunk 4 above)
Comment (zazwaz12) [reply 1. extraction/gnaf.py:45]: To have an Excel reference for the code meanings.

File: 1. extraction/gnaf.py (hunk 5 above)
Comment (zazwaz12) [reply 1. extraction/gnaf.py:62]: Is redundant in both, will remove.

File: syncdata.py
Comment (zazwaz12) [reply syncdata.py:1]: Mistake.

File: 1. extraction/gnaf.py (hunk 10 above)
Comment (zazwaz12) [reply 1. extraction/gnaf.py:1]: You can implement Polars as long as it works, Pandas works.

File: 2. transformation/houses.py (hunk 11 above)
Comment (zazwaz12) [reply 2. transformation/houses.py:16]: Happy to make the global variables a config going forward, but for this it is a very simple script that will not change or increase in complexity.

**Comment by SodaVolcano** [comment]
Closed by #36
